### PR TITLE
Optional resource attribute enforcement

### LINKED ISSUE #81: Support for optional resource attribute validation
Currently, server's authorization of OAuth request [returns an error response](https://github.com/IdentityPython/idpy-oidc/blob/869068cd36ee073aae5e23b4b40a1c97a5cec5ca/src/idpyoidc/server/oauth2/authorization.py#L285C7-L285C7) 
 when `resource_indicators` policy is configured and `resource` attribute is not provided in the request. However, `resource` indicator is an optional attribute, therefore some of the incoming requests might contain it while others may not. 

We'd like to accommodate both of these possibilities, validate resource indicators when they are provided and skip the process in case they're not present to prevent the error message for clients not specifying the `resource` indicator in their request. One such possibility might be to add a wrapper method for [validate_resource_indicators_policy](https://github.com/IdentityPython/idpy-oidc/blob/869068cd36ee073aae5e23b4b40a1c97a5cec5ca/src/idpyoidc/server/oauth2/authorization.py#L284C30-L284C30) which would not initiate the validation if the `resource` attribute is missing. Something like this:

```
def optional_validate_resource_indicators_policy(request, context, **kwargs):
    if "resource" not in request:
        return request

    return validate_resource_indicators_policy(request, context, **kwargs)
```

No other changes would be required. This method could then simply be configured in resource_indicators config like so:
```
resource_indicators:
   policy:
       function: idpyoidc.server.oauth2.authorization.optional_validate_resource_indicators_policy
```

Edit: I've provided a [PR](https://github.com/IdentityPython/idpy-oidc/pull/82) from my fork.

### PR DESCRIPTION
**Pull request by PeterBolha**
Fixes #81

Provides a wrapper method that does not return an error when the validate resource attribute policy is set but the attribute is not provided in the client's request.

### CODE REVIEW

**Comment by c00kiemon5ter** [comment]
Please, point the PR towards the `main` branch

**Comment by ctriant** [comment]
Hey @PeterBolha thanks for the contribution! 
I think it would be a good idea to replace the check regarding the existence of `resource` in `request` of the original `validate_resource_indicators_policy` with the one you propose on your `optional_validate_resource_indicators_policy`.

Thus, instead of returning 
`oauth2.AuthorizationErrorResponse(
            error="invalid_target",
            error_description="Missing resource parameter",
)` we will return the actual request.

That way we'll not need the optional wrapper, and therefore a mechanism/switch to enable it.

@c00kiemon5ter does this sound reasonable?

**Comment by c00kiemon5ter** [comment]
Yes, it makes sense. This is what the RFC specifies

> In requests to the authorization server, a client **MAY indicate the protected resource** (a.k.a. resource server, application, API, etc.) to which it is requesting access by including the following parameter in the request. [...]

emphasis on MAY, meaning that requests without the `resource` query param should still be processed.

**Review by ctriant — APPROVED** [review]

**Review by ctriant — DISMISSED** [review]
Fixes will be needed on tests after this change.

**Comment by PeterBolha** [comment]
I have fixed the related test, however there are two more failing ones. I've rebased my branch to have the latest `main` branch. Just now I'm noticing that the latest main has these two errors popping up in tests. Should I wait for the fixes in main and rebase later or figure out something else?

**Comment by ctriant** [comment]
> I have fixed the related test, however there are two more failing ones. I've rebased my branch to have the latest `main` branch. Just now I'm noticing that the latest main has these two errors popping up in tests. Should I wait for the fixes in main and rebase later or figure out something else?

Thanks! Don't worry about these tests. Let's wait a bit before we merge this, and if necessary I'll ask you to rebase your branch later on.

**Review by c00kiemon5ter — APPROVED** [review]

**Pull request merged**
